odoo/addons/base/models/ir_qweb.py

```diff
@@ -515,52 +517,102 @@ def keep_query(*keep_params, **additional_params):
+    def __init__(self, message, qweb, template=None, ref=None, path_xml=None, code=None, path_info=None):
+        self.title = message
         self.stack = traceback.format_exc()
-        self.name = template
-        self.ref = ref
-        self.path, self.html = path_xml or (None, None)
-        self.code = None
+        self.name = ref if template is None else template
+        self.ref, self.path, self.html = path_xml or (ref, None, None)
+        self.path_info = path_info
+        self.code = code if qweb.env.context.get('dev_mode') else None
         if code:
-            self.code = '\n'.join(code.split('\n')[:-1]) if qweb.env.context.get('dev_mode') else None
             line_nb = 0
             for error_line in reversed(self.stack.split('\n')):
-                if f'File "<{self.ref}>"' in error_line:
+                if f'File "<{self.ref}>"' in error_line or (self.ref is None and 'File "<' in error_line):
                     line_function = error_line.split(', line ')[1]
                     line_nb = int(line_function.split(',')[0])
                     break
             for code_line in reversed(code.split('\n')[:line_nb]):
+                if code_line.startswith('def '):
+                    break
                 match = re.match(r'\s*# element: (.*) , (.*)', code_line)
                 if match:
                     self.path = match[1][1:-1]
                     self.html = match[2][1:-1]
                     break
 
-        self.title = message
         super().__init__(message)
 
-    def __str__(self):
-        parts = [self.title]
-        if self.__cause__ and str(self.__cause__) != '':
-            parts.append(f"{self.__cause__.__class__.__name__}: {self.__cause__}")
-        elif self.__context__ and str(self.__context__) != '':
-            parts.append(f"{self.__context__.__class__.__name__}: {self.__context__}")
+    def __init_subclass__(cls):
+        if QWebException in cls.__bases__:
+            QWEB_ERROR_CLASSES[cls.__bases__[1]] = cls
+        return super().__init_subclass__()
+
+    @property
+    def detail(self):
+        parts = []
         if self.name is not None:
             parts.append(f"Template: {self.name}")
+        if self.ref is not None:
+            parts.append(f"Reference: {self.ref}")
         if self.path is not None:
             parts.append(f"Path: {self.path}")
         if self.html is not None:
-            parts.append(f"Node: {self.html}")
+            parts.append(f"Element: {self.html}")
+        if self.path_info is not None:
+            path_info = '\n      '.join(str(v) for v in self.path_info)
+            if path_info:
+                parts.append(f"From: {path_info}")
         if self.code is not None:
-            parts.append(f"Compiled code:\n{self.code}")
+            parts.append(f"Compiled code:\n\n{self.code}")
         return "\n".join(parts)
 
+    def __str__(self):
+        errorType = "Exception"
+        if str(self.__cause__):
```

Comment:
```suggestion
        if self.__cause__ is not None:
```

Because `str(None) == 'None'` which is truthy.
Same just below.